File: app/src/main/java/my/ew/wallpaper/utils/AnimUtils.java
package my.ew.wallpaper.utils;

import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.LinearLayout;

/**
 * Created by isanechek on 09.08.15.
 */
public class AnimUtils extends Animation {

    public final static int COLLAPSED = 1;
    public final static int EXPAND = 0;

    private View vv;
    private int endHeight;
    private int mType;
    private LinearLayout.LayoutParams lp;

    public AnimUtils(View view, int duration, int type) {
        setDuration(duration);
        vv = view;
        endHeight = vv.getHeight();
        lp = ((LinearLayout.LayoutParams) vv.getLayoutParams());
        mType = type;
        if (mType == EXPAND) {
            lp.height = 0;
        } else {
            lp.height = ViewGroup.LayoutParams.WRAP_CONTENT;
        }
        vv.setVisibility(View.VISIBLE);
    }

    public int getHeight() {
        return vv.getHeight();
    }

    public void setHeight(int height) {
        endHeight = height;
    }

    @Override
    protected void applyTransformation(float interpolatedTime, Transformation t) {
        super.applyTransformation(interpolatedTime, t);
        if (interpolatedTime < 1.0f) {
            if (mType == EXPAND) {
                lp.height = (int)(endHeight * interpolatedTime);
            } else {
                lp.height = (int)(endHeight * (1 - interpolatedTime));
            }
            vv.requestLayout();
        } else {
            if (mType == EXPAND) {
                lp.height = ViewGroup.LayoutParams.WRAP_CONTENT;
                vv.requestLayout();
            } else {
                vv.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/my/ew/wallpaper/utils/AnimUtils.kt
package my.ew.wallpaper.utils

import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.Transformation
import android.widget.LinearLayout

/**
 * Created by isanechek on 09.08.15.
 */
class AnimUtils(val vv: View, duration: Int, val mType: Int) : Animation() {
    private var endHeight: Int = 0
    private val lp: LinearLayout.LayoutParams

    init {
        setDuration(duration.toLong())
        endHeight = vv.height
        lp = (vv.layoutParams as LinearLayout.LayoutParams)
        if (mType == EXPAND) {
            lp.height = 0
        } else {
            lp.height = ViewGroup.LayoutParams.WRAP_CONTENT
        }
        vv.visibility = View.VISIBLE
    }

    fun getHeight(): Int {
        return vv.height
    }

    fun setHeight(height: Int) {
        endHeight = height
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
        super.applyTransformation(interpolatedTime, t)
        if (interpolatedTime < 1.0f) {
            if (mType == EXPAND) {
                lp.height = (endHeight * interpolatedTime).toInt()
            } else {
                lp.height = (endHeight * (1 - interpolatedTime)).toInt()
            }
            vv.requestLayout()
        } else {
            if (mType == EXPAND) {
                lp.height = ViewGroup.LayoutParams.WRAP_CONTENT
                vv.requestLayout()
            } else {
                vv.visibility = View.GONE
            }
        }
    }

    companion object {
        const val COLLAPSED: Int = 1
        const val EXPAND: Int = 0
    }
}
